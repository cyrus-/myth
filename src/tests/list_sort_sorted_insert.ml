let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

type cmp =
  | LT
  | EQ
  | GT

let rec compare (n1 : nat) (n2 :nat) : cmp =
  match n1 with
  | O -> ( match n2 with
           | O -> EQ
           | S (m) -> LT
         )
  | S (m1) ->
      ( match n2 with
      | O -> GT
      | S (m2) -> (compare m1 m2) )
;;

let rec insert (l : list) (n :nat) : list =
  match l with
  | Nil -> Cons(n, Nil)
  | Cons(m, tl) ->
    (match compare n m with
     | LT -> Cons (n, Cons(m, tl))
     | EQ -> l
     | GT -> Cons (m, insert tl n)
    )
;;

let list_sort_sorted_insert : list -> list |>
 { [] => []
 | [0] => [0]
 | [1] => [1]
 | [0;0] => [0]
 | [1;0] => [0;1]
 | [1;1] => [1]
 | [0;1;1] => [0;1]
 } = ?
"
