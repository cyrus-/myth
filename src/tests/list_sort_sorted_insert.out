(* Generated Thu Nov 13 00:35:51 2014 *)
let list_sort_sorted_insert : list -> list =
  let rec f1 (l1:list) : list =
    match l1 with
      | Nil -> Nil
      | Cons (n1, l2) -> insert (f1 l2) n1
  in
    f1
;;
